lib/guardian/plug/ensure_authenticated.ex

```diff
@@ -1,79 +1,67 @@
 
-  ## Example
+    This, like all other Guardian plugs, requires a Guardian pipleine to be setup.
+    It requires an implementation module, an error handler and a key.
 
-      # Will call the unauthenticated/2 function on your handler
-      plug Guardian.Plug.EnsureAuthenticated, handler: SomeModule
+    These can be set either:
 
-      # look in the :secret location.  You can also do simple claim checks:
-      plug Guardian.Plug.EnsureAuthenticated, handler: SomeModule, key: :secret
+    1. Upstream on the connection with `plug Guardian.Pipeline`
+    2. Upstream on the connection with `Guardian.Pipeline.{put_module, put_error_handler, put_key}`
+    3. Inline with an option of `:module`, `:erorr_handler`, `:key`
 
-      plug Guardian.Plug.EnsureAuthenticated, handler: SomeModule, typ: "access"
+    Options:
 
-  If the handler option is not passed, `Guardian.Plug.ErrorHandler` will provide
-  the default behavior.
-  """
-  require Logger
-  import Plug.Conn
+    * `claims` - The literal claims to check to ensure that a token is valid
+    * `key` - The location to find the information in the connection. Defaults to: `deafult`
 
-  @doc false
-  def init(opts) do
-    opts = Enum.into(opts, %{})
-    handler = build_handler_tuple(opts)
+    ## Example
 
-    claims_to_check = Map.drop(opts, [:on_failure, :key, :handler])
-    %{
-      handler: handler,
-      key: Map.get(opts, :key, :default),
-      claims: Guardian.Utils.stringify_keys(claims_to_check)
-    }
-  end
+    ```elixir
 
-  @doc false
-  def call(conn, opts) do
-    key = Map.get(opts, :key, :default)
+      # setup the upstream pipeline
+      plug Guardian.Plug.EnsureAuthenticated, claims: %{"typ" => "access"}
+      plug Guardian.Plug.EnsureAuthenticated, key: :secret
+      ```
+    """
+    alias Guardian.Plug, as: GPlug
+    alias GPlug.{Pipeline}
+    alias Guardian.Token.Verify
 
-    case Guardian.Plug.claims(conn, key) do
-      {:ok, claims} -> conn |> check_claims(opts, claims)
-      {:error, reason} -> handle_error(conn, {:error, reason}, opts)
-    end
-  end
+    import Plug.Conn
 
-  defp handle_error(%Plug.Conn{params: params} = conn, reason, opts) do
-    conn = conn |> assign(:guardian_failure, reason) |> halt
-    params = Map.merge(params, %{reason: reason})
-    {mod, meth} = Map.get(opts, :handler)
-
-    apply(mod, meth, [conn, params])
-  end
+    @doc false
+    def init(opts), do: opts
 
-  defp check_claims(conn, opts = %{claims: claims_to_check}, claims) do
-    claims_match =
-      claims_to_check
-      |> Map.keys
-      |> Enum.all?(&(claims_to_check[&1] == claims[&1]))
+    @doc false
+    def call(conn, opts) do
+      token = GPlug.current_token(conn, opts)
+      claims = GPlug.current_claims(conn, opts)
 
-    if claims_match do
-      conn
-    else
-      handle_error(conn, {:error, :claims_do_not_match}, opts)
+      if token do
+        case verify_claims(claims, opts) do
```

Comment:
We could refactor this to DRY-up the error handling and remove the nested if/case:

```elixir
def call(conn, opts) do
  conn
  |> GPlug.current_token(opts)
  |> verify
  |> respond
end

defp verify(nil, _opts), do: {:error, :unauthenticated}
defp verify(_token, opts) do
  conn
  |> GPlug.current_claims(opts)
  |> verify_claims(opts)
end

def respond({:ok, _}, conn, _opts), do: conn
def respond({:error, reason}, conn, opts) do
  conn
  |> Pipeline.fetch_error_handler!(opts)
  |> apply(:auth_error, [conn, {:unauthenticated, reason}, opts])
  |> halt()
end
```